graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/tracing/FederatedTracingInstrumentation.java

```diff
@@ -0,0 +1,282 @@
+import graphql.execution.ExecutionStepInfo;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldFetchParameters;
+import graphql.schema.DataFetchingEnvironment;
+import mdg.engine.proto.Reports;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+import static graphql.execution.instrumentation.SimpleInstrumentationContext.whenCompleted;
+import static graphql.schema.GraphQLTypeUtil.simplePrint;
+
+public class FederatedTracingInstrumentation extends SimpleInstrumentation {
+    private static final String KEY = "ftv1";
+    private static final String DEBUG_KEY = "ftv1_debug";
+
+    private boolean debuggingEnabled = false;
+
+    public FederatedTracingInstrumentation() {
+    }
+
+    public FederatedTracingInstrumentation(boolean debug) {
+        debuggingEnabled = debug;
+    }
+
+    @Override
+    public InstrumentationState createState() {
+        return new FederatedTracingState();
+    }
+
+    @Override
+    public CompletableFuture<ExecutionResult> instrumentExecutionResult(ExecutionResult executionResult, InstrumentationExecutionParameters parameters) {
+        FederatedTracingState state = parameters.getInstrumentationState();
+
+        Map<Object, Object> extensions = executionResult.getExtensions();
+        Map<Object, Object> tracingMap = new LinkedHashMap<>(extensions == null ? Collections.emptyMap() : extensions);
+
+        Reports.Trace trace = state.toProto();
+
+        byte[] bytes = trace.toByteArray();
+        tracingMap.put(KEY, Base64.getEncoder().encodeToString(bytes));
+
+        if (debuggingEnabled) {
+            tracingMap.put(DEBUG_KEY, trace.toString());
+        }
+
+        return CompletableFuture.completedFuture(new ExecutionResultImpl(executionResult.getData(), executionResult.getErrors(), tracingMap));
+    }
+
+    @Override
+    public InstrumentationContext<Object> beginFieldFetch(InstrumentationFieldFetchParameters parameters) {
+        FederatedTracingState state = parameters.getInstrumentationState();
+
+        long startNanos = System.nanoTime();
+        return whenCompleted((result, throwable) -> {
+            long endNanos = System.nanoTime();
+
+            DataFetchingEnvironment env = parameters.getEnvironment();
+            ExecutionStepInfo executionStepInfo = env.getExecutionStepInfo();
+
+            Node node = state.addNode(executionStepInfo, startNanos, endNanos);
+
+            if (throwable != null) {
+                node.addError(throwable);
```

Comment:
feat: It might be useful to allow implementors to decide how they want to translate thrown errors into GraphQL errors that services report to the gateway, but we can definitely do that when we actually have a feature request and not during this round of code review 😉

Reply:
Yeah, I left that feature out of the ruby implementation too. One thing I might do today is make it easier to add configuration options for tracing, which would give us a place to add a error translation hook.